### PR TITLE
timeloop: init at unstable-2022-10-28 and accelergy: init at unstable-2022-05-03

### PR DESCRIPTION
**Pull request by gdinh**
###### Description of changes

Added Timeloop and Accelergy, two tools to model energy and area for accelerator chips. Timeloop depends on accelergy by default, as it requires an energy model to work (the [upstream documentation](https://timeloop.csail.mit.edu/timeloop/installation) assumes they will be installed together). However, this can be disabled by setting the enableAccelergy input to false if a user has their own energy model.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hqjenny** [comment]
I'm a developer of projects relying on the Timeloop project, it would be nice to have this package added to nix.

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    #gpm doesn't exist on darwin
+    substituteInPlace ./src/SConscript --replace ", 'gpm'" ""
+   '';
+
+  sconsFlags =
+    # will fail on clang/darwin on link without --static due to undefined extern
+    # however, will fail with static on linux as nixpkgs deps aren't static
+    (lib.optional stdenv.isDarwin "--static")
+    ++ (lib.optional enableAccelergy "--accelergy")
+    ++ (lib.optional enableISL "--with-isl");
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:79]:
```suggestion
  sconsFlags =
    # will fail on clang/darwin on link without --static due to undefined extern
    # however, will fail with static on linux as nixpkgs deps aren't static
    lib.optional stdenv.isDarwin "--static"
    ++ lib.optional enableAccelergy "--accelergy"
    ++ lib.optional enableISL "--with-isl";
```

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+
+  installPhase = ''
+    cp -r ./bin $out
+    cp -r ./lib $out
+    cp -r ./problem-shapes $out
+    cp -r ./doc $out
+    cp -r ./configs $out
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:87]:
```suggestion
    cp -r ./bin ./lib ./problem-shapes ./doc ./configs $out
```
problem-shapes and configs should go into a subdirectory. doc should go under share

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    #disable LTO on macos
+    substituteInPlace ./src/SConscript --replace ", '-flto'" ""
+
+    #static builds on mac fail as no static libcrt is provided by apple
+    #see https://stackoverflow.com/questions/3801011/ld-library-not-found-for-lcrt0-o-on-osx-10-6-with-gcc-clang-static-flag
+    substituteInPlace ./src/SConscript --replace "'-static-libgcc', " ""
+    substituteInPlace ./src/SConscript --replace "'-static-libstdc++', " ""
+    substituteInPlace ./src/SConscript --replace "'-Wl,--whole-archive', '-static', " ""
+    substituteInPlace ./src/SConscript --replace ", '-Wl,--no-whole-archive'" ""
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:64]:
```suggestion
    # static builds on mac fail as no static libcrt is provided by apple
    # see https://stackoverflow.com/questions/3801011/ld-library-not-found-for-lcrt0-o-on-osx-10-6-with-gcc-clang-static-flag
    substituteInPlace ./src/SConscript \
      --replace "'-static-libgcc', " "" \
      --replace "'-static-libstdc++', " "" \
      --replace "'-Wl,--whole-archive', '-static', " "" \
      --replace ", '-Wl,--no-whole-archive'" ""
```

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  patchPhase = ''
+    #use nix ar/ranlib
+    substituteInPlace ./SConstruct --replace "env.Replace(AR = \"gcc-ar\")" ""
+    substituteInPlace ./SConstruct --replace "env.Replace(RANLIB = \"gcc-ranlib\")" ""
+    '' + lib.optionalString stdenv.isDarwin ''
+    #prevent clang from dying on errors that gcc is fine with
+    substituteInPlace ./src/SConscript --replace "-Werror" "-Wno-inconsistent-missing-override"
+
+    #disable LTO on macos
+    substituteInPlace ./src/SConscript --replace ", '-flto'" ""
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:57]:
```suggestion
    # use nix ar/ranlib
    substituteInPlace ./SConstruct \
      --replace "env.Replace(AR = \"gcc-ar\")" ""
      --replace "env.Replace(RANLIB = \"gcc-ranlib\")" ""
    '' + lib.optionalString stdenv.isDarwin ''
    # prevent clang from dying on errors that gcc is fine with
    substituteInPlace ./src/SConscript --replace "-Werror" "-Wno-inconsistent-missing-override"

    # disable LTO on macos
    substituteInPlace ./src/SConscript --replace ", '-flto'" ""
```
Please use a space after the hashtag everywhere

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+     cp -r ./pat-public/src/pat ./src/pat
+   '';
+
+  enableParallelBuilding = true;
+
+  #link-time optimization fails on darwin
+  #see https://github.com/NixOS/nixpkgs/issues/19098
+  NIX_CFLAGS_COMPILE = lib.optional stdenv.isDarwin "-fno-lto";
+
+  patchPhase = ''
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:48]:
```suggestion
  postPatch = ''
```

File: pkgs/applications/science/computer-architecture/timeloop/default.nix [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    boost
+    libyaml
+    libyamlcpp
+    ncurses
+    accelergy
+   ] ++ lib.optionals stdenv.isLinux [ gpm ];
+
+  preConfigure = ''
+     cp -r ./pat-public/src/pat ./src/pat
+   '';
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/timeloop/default.nix:40]:
```suggestion
  preConfigure = ''
    cp -r ./pat-public/src/pat ./src/pat
  '';
```

File: pkgs/applications/science/computer-architecture/accelergy/default.nix [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    pyyaml
+    yamlordereddictloader
+    pyfiglet
+    setuptools
+  ];
+
+  meta = with lib; {
+    description = "An architecture-level energy/area estimator for accelerator designs";
+    license = licenses.mit;
+    homepage = "http://accelergy.mit.edu";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/accelergy/default.nix:26]:
```suggestion
    homepage = "https://accelergy.mit.edu/";
```

File: pkgs/applications/science/computer-architecture/accelergy/default.nix [hunk 8]
```diff
@@ -0,0 +1,29 @@
+{ lib, fetchFromGitHub, python3Packages, pkgs }:
+
+python3Packages.buildPythonApplication rec {
+  pname = "accelergy";
+  version = "unstable-2022-05-03";
+
+  disabled = !python3Packages.isPy3k;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/science/computer-architecture/accelergy/default.nix:8]:
```suggestion
```
python3Packages cannot not be python3, it is never python 2.7.

**Comment by gdinh** [comment]
Just did a version bump, no other changes to this PR since Nov 26.

**Pull request merged**